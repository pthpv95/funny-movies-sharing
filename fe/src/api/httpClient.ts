type ResponseType = 'json' | 'text' | 'blob';
const baseUrl = 'http://localhost:3001';

export class HttpClient {
  private getUrl(relativeUrl: string) {
    return `${baseUrl}${relativeUrl}`;
  }

  public get(
    relativeUrl: string,
    headers = {},
    responseType: ResponseType = 'json'
  ) {
    return fetch(this.getUrl(relativeUrl), {
      credentials: 'include',
      headers: {
        ...headers,
        'Content-Type': 'application/json',
      },
    }).then((res) => this.status(res, responseType));
  }

  public post(
    relativeUrl: string,
    body: object,
    headers = {},
    responseType: ResponseType = 'json'
  ) {
    return fetch(this.getUrl(relativeUrl), {
      body: JSON.stringify(body),
      credentials: 'include',
      headers: {
        ...headers,
        'Content-Type': 'application/json',
      },
      method: 'POST',
    }).then((res) => this.status(res, responseType));
  }

  private status(res: any, responseType: ResponseType) {
    if (!res.ok) {
      throw res
    }
    if (responseType === 'text') {
      return res.text();
    }

    if (responseType === 'blob') {
      return res.blob();
    }

    return res.json();
  }
}

export const httpClient = new HttpClient();
