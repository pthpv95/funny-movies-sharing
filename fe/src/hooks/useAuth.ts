import React, { useContext } from "react";

export let AuthContext = React.createContext<
  {
    authUser: any,
    signIn: (email: string, password: string) => Promise<void>,
    signOut: () => void
  }>({
    authUser: null,
    signIn: async (email: string, password: string) => { },
    signOut: () => { },
  });

let useAuth = () => useContext(AuthContext)

export default useAuth;