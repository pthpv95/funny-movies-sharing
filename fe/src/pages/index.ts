export { default as Home } from './Home';
export { default as ShareMovie } from './ShareMovie';
export { default as Layout } from './Layout';
