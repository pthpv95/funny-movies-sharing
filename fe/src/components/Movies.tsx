import React, { useEffect, useState } from 'react';
import { httpClient } from '../api/httpClient';

interface IMovie {
  _id: string;
  title: string;
  url: string;
  description: string;
  embedId: string;
  createdBy: string;
}

const MovieItem = ({ item }: { item: IMovie }) => {
  return (
    <div className="movie-item">
      <div>
        <iframe
          width={450}
          height={300}
          src={`https://www.youtube.com/embed/${item.embedId}`}
          title={item.title}
          frameBorder="0"
          allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
          allowFullScreen
        ></iframe>
      </div>
      <div className="info">
        <p className="title">{item.title}</p>
        <p><strong>Shared by: {item.createdBy}</strong></p>
        <p><strong>Description: </strong></p>
        <div className="description">{item.description}</div>
      </div>
    </div>
  );
};

const Movies = () => {
  let [movies, setMovies] = useState<IMovie[]>();
  useEffect(() => {
    httpClient.get('/api/movies').then((res) => {
      setMovies(res.data.movies);
    });
  }, []);

  return (
    <div className="movie-list">
      {movies?.map((item) => {
        return <MovieItem key={item._id} item={item} />;
      })}
    </div>
  );
};

export default Movies;
