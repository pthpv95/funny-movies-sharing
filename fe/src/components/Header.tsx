import { useAuth } from '../hooks';
import SignInUpForm from './SignInUp';
import { useNavigate, Link } from 'react-router-dom';

let LoggedIn = () => {
  const { authUser, signOut } = useAuth();
  let navigate = useNavigate();
  return (
    <div className="logged-in-state">
      <p>Welcome {authUser.email}</p>
      <button
        onClick={(e) => {
          e.preventDefault();
          navigate('/share');
        }}
      >
        Share a movie
      </button>
      <button onClick={signOut}>Logout</button>
    </div>
  );
};

const Header = () => {
  const { authUser } = useAuth();
  return (
    <div className="header">
      <Link to="/" style={{ textDecoration: 'none' }}>
        <h2>🏠Funny Movies</h2>
      </Link>
      {authUser ? <LoggedIn /> : <SignInUpForm />}
    </div>
  );
};

export default Header;
