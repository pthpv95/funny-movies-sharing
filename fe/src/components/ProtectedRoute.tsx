import { Navigate, Outlet } from 'react-router-dom';
import { useAuth } from "../hooks";

const ProtectedRoute = () => {
  const { authUser } = useAuth();

  if (!authUser) {
    return <Navigate to="/" />
  }

  return <Outlet />;
};

export default ProtectedRoute;