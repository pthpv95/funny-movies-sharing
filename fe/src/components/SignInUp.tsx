import { useState } from 'react';
import { useAuth } from '../hooks';

let SignInUp = () => {
  let [email, setEmail] = useState('');
  let [password, setPassword] = useState('');
  let { signIn } = useAuth();

  let handleSubmit = async (e: any) => {
    e.preventDefault();
    await signIn(email, password);
  };

  return (
    <form onSubmit={handleSubmit} className="form-sign-in-up">
      <input
        type="email"
        value={email}
        required
        placeholder="email"
        onChange={(e: any) => setEmail(e.target.value)}
      />
      <input
        type="password"
        value={password}
        required
        placeholder="password"
        onChange={(e: any) => setPassword(e.target.value)}
      />
      <button type="submit">Login / Register</button>
    </form>
  );
};

export default SignInUp;
