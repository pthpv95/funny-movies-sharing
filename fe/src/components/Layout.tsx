import Header from './Header';
import { Outlet } from 'react-router-dom';
let Layout = () => {
  return (
    <div className="App">
      <Header />
      <Outlet />
    </div>
  );
};

export default Layout;
