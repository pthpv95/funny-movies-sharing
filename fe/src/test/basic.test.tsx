import { create } from 'react-test-renderer';
import { describe, expect, it, vi } from 'vitest';
import { Home, Layout, ShareMovie } from '../pages';

import { MemoryRouter, Route, Routes } from 'react-router-dom';
import { httpClient } from '../api/httpClient';
import { Header, ProtectedRoute } from '../components';
import { AuthContext } from '../hooks/useAuth';

describe('App', () => {
  it('renders correctly', () => {
    vi.spyOn(httpClient, 'get').mockResolvedValue({ data: { movies: [] } });
    let renderer = create(
      <MemoryRouter initialEntries={['/']}>
        <Routes>
          <Route path="/" element={<Layout />}>
            <Route path="/" element={<Home />} />
            <Route element={<ProtectedRoute />}>
              <Route path="/share" element={<ShareMovie />} />
            </Route>
          </Route>
        </Routes>
      </MemoryRouter>
    );

    expect(renderer.toJSON()).toMatchSnapshot();
  });
});

describe('Header', () => {
  describe('Given view for guest', () => {
    it('should render sign in/up', () => {
      let renderer = create(
        <MemoryRouter>
          <Header />
        </MemoryRouter>
      );
      expect(renderer.toJSON()).toMatchSnapshot();
    });
  });

  describe('Given view for signed in user', () => {
    it('should render Logout button', () => {
      let renderer = create(
        <AuthContext.Provider
          value={{
            authUser: { email: 'test@gmail.com' },
            signOut: () => {},
            signIn: async (email: string, password: string) => {},
          }}
        >
          <MemoryRouter>
            <Header />
          </MemoryRouter>
        </AuthContext.Provider>
      );
      let snapshot: any = renderer.toJSON();
      let loggedInSnapshot = snapshot.children.find((c: any) => c.type === 'div');
      let shareMovieButton = loggedInSnapshot.children.find((c: any) => c.type === 'button')
      
      expect(shareMovieButton.children[0]).toEqual('Share a movie');
      expect(loggedInSnapshot.props.className).toEqual('logged-in-state')
      expect(renderer.toJSON()).toMatchSnapshot();
    });
  });
});
