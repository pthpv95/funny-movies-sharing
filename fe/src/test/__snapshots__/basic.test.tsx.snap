// Vitest Snapshot v1

exports[`App > renders correctly 1`] = `
<div
  className="App"
>
  <div
    className="header"
  >
    <a
      href="/"
      onClick={[Function]}
      style={
        {
          "textDecoration": "none",
        }
      }
    >
      <h2>
        🏠Funny Movies
      </h2>
    </a>
    <form
      className="form-sign-in-up"
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        placeholder="email"
        required={true}
        type="email"
        value=""
      />
      <input
        onChange={[Function]}
        placeholder="password"
        required={true}
        type="password"
        value=""
      />
      <button
        type="submit"
      >
        Login / Register
      </button>
    </form>
  </div>
  <div
    className="movie-list"
  />
</div>
`;

exports[`Header > Given view for guest > should render sign in/up 1`] = `
<div
  className="header"
>
  <a
    href="/"
    onClick={[Function]}
    style={
      {
        "textDecoration": "none",
      }
    }
  >
    <h2>
      🏠Funny Movies
    </h2>
  </a>
  <form
    className="form-sign-in-up"
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      placeholder="email"
      required={true}
      type="email"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="password"
      required={true}
      type="password"
      value=""
    />
    <button
      type="submit"
    >
      Login / Register
    </button>
  </form>
</div>
`;

exports[`Header > Given view for signed in user > should render Logout button 1`] = `
<div
  className="header"
>
  <a
    href="/"
    onClick={[Function]}
    style={
      {
        "textDecoration": "none",
      }
    }
  >
    <h2>
      🏠Funny Movies
    </h2>
  </a>
  <div
    className="logged-in-state"
  >
    <p>
      Welcome 
      test@gmail.com
    </p>
    <button
      onClick={[Function]}
    >
      Share a movie
    </button>
    <button
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
</div>
`;
